lib/mongoid/matcher.rb

```diff
@@ -50,8 +47,9 @@ module Matcher
         current.each do |doc|
           case doc
           when Hash
-            if doc.key?(field)
-              new << doc[field]
+            v = indifferent_hash_fetch(doc, field)
+            if !v.nil?
```

Comment:
this change is treating `does not have key` the same as `has key with nil value`,
and as such breaks querying with `$exists` (at least):

```console
$ bin/rails console
class Model
  include Mongoid::Document

  field :a
  field :b
end

m = Model.new(a:1, b:nil)
m.send(:as_attributes)
=> {"_id"=>BSON::ObjectId('60bf7ad9fb97faffc6b460ec'), "a"=>1, "b"=>nil}

# --------------------
# before these changes, $exists correctly returns 'true'

m._matches?(:b => {:"$exists" => true})
=> true

# --------------------
# with these changes installed, $exists now returns 'false' incorrectly

m._matches?(:b => {:"$exists" => true})
=> false
```

before this PR, extract_attribute was returning `[nil]` and is now returning `[]`

Reply:
Nice catch! I'm making a fix for this now and will have a PR up soon.